README.md
gogrs [![Release](https://img.shields.io/github/release/toomore/gogrs.svg)](https://github.com/toomore/gogrs/releases)
======

[![GoDoc](https://godoc.org/github.com/toomore/gogrs?status.svg)](https://godoc.org/github.com/toomore/gogrs) [![Coverage Status](https://coveralls.io/repos/toomore/gogrs/badge.svg?branch=master)](https://coveralls.io/r/toomore/gogrs?branch=master) [![GitHub license](https://img.shields.io/badge/license-MIT-blue.svg)](https://raw.githubusercontent.com/toomore/gogrs/master/LICENSE) [![Build Status](https://travis-ci.org/toomore/gogrs.svg?branch=master)](https://travis-ci.org/toomore/gogrs) [![Go Report Card](https://goreportcard.com/badge/github.com/toomore/gogrs)](https://goreportcard.com/report/github.com/toomore/gogrs)

gogrs is a tool for fetching data from Taiwan Stock Exchange(TWSE) and dockerizing. gogrs now is still in development. I will try my best to speed up to completed the same function with [grs](https://github.com/toomore/grs) (Python). gogrs 是擷取台灣上市股票股價資訊工具，目前還在大量的開發中。原始工具是用 [grs](https://github.com/toomore/grs)（Python 套件），目標是將基本功能用 go 來實作。

Packages
---------

1. realtime - [擷取盤中個股、指數即時股價資訊](https://godoc.org/github.com/toomore/gogrs/realtime) **（使用太過頻繁會有被擋掉的風險，目前無完善的解決辦法）**
2. twse - [擷取台灣股市上市/上櫃股票資訊、上市/上櫃類股清單、外資及陸資持股比率前二十名彙總表、 三大法人買賣金額統計表、三大法人買賣超日報、自營商、投信、外資及陸資買賣超彙總表](https://godoc.org/github.com/toomore/gogrs/twse)
3. tradingdays - [股市開休市判斷（支援非國定假日：颱風假）與當日區間判斷（盤中、盤後、盤後盤）](https://godoc.org/github.com/toomore/gogrs/tradingdays)
4. utils - [套件所需的公用工具（總和、平均、序列差、持續天數、民國日期解析、簡單亂數、標準差、簡單 net/http 快取）](https://godoc.org/github.com/toomore/gogrs/utils)

Cmd
----

![gogrs cmd demo](https://s3-ap-northeast-1.amazonaws.com/toomore/gogrs/gogrs_cmd_demo_20150615.png "gogrs cmd demo")

1. example - [簡單範例測試](cmd/docs/gogrs_example.md)
2. realtime - [擷取盤中即時資訊與大盤、上櫃、寶島指數](cmd/docs/gogrs_realtime.md)
3. report - [每日收盤後產生符合選股條件的報告](cmd/docs/gogrs_report.md)
4. cache - [清除 twse cache](cmd/docs/gogrs_cache.md)
5. server - [提供簡單的日期查詢 API Server](cmd/docs/gogrs_server.md)

相關的操作請參考 `-h` 的說明，或 [cmd/docs](cmd/docs/gogrs.md)

Docker
-------

Download image.

    docker pull toomore/gogrs

`tag:latest` bind to `branch:master`, more docker [info](https://registry.hub.docker.com/u/toomore/gogrs/).

Or ...

Build minify gogrs docker(`toomore/gogrs:mini`)

    cd ./build-mini.sh

Run `tradingdays server`.

    docker run -d -p 80:59123 toomore/gogrs:mini gogrs server

Or login run other cmd

    docker run -it toomore/gogrs:mini

Create a ramdisk volume

    docker create -v /run/shm/:/run/shm --name ramdisk toomore/gogrs:mini

Run with ramdisk volume

    docker run -it --volumes-from ramdisk toomore/gogrs:mini

TODO
-----

1. In English comment.
2. 盤中預估量能。
3. 個股對應分類股資訊。
4. 個股融資融券資訊。
5. 顯示三大法人。
6. 各類股盤後買賣超表。
7. 新聞內容擷取。

Known Issue
------------

1. None.

## Changelog

### v1.4 (2017/07/08)
* cmd using spf13/cobra

### v1.3 (2017/07/07)
* Fixed twse change source link. #4

### v1.2

#### v1.2.0 (2017/04/25)
* Fixed `realtime` #1 #3

#### v1.2.1 (2017/04/26)
* Pass realtime testing
* Put gogrs:mini into test

### v1.1

#### v1.1.0 (2017/04/07)
* Fixed `GetOSRamdiskPath`
* Add `Weight`, `WeightVolume`
* Add more test case

#### v1.1.1 (2017/04/08)
* Fixed Dockerfile

#### v1.1.2 (2017/04/08)
* Docker image using `alpine`
* Fixed image:`gogrs:mini` (`sh ./docker-mini/make.sh`)

#### v1.1.3 (2017/04/10)
* Fixed Dockerfile
* Update open time 2011 ~ 2017

### v1.0.0 (2016/10/08)
* Stable release for dependency tools.

### v0.1.0 (2016/10/08)
* gogrs first package version.

License
--------

The MIT License (MIT)

Copyright © 2015, 2016, 2017 Toomore Chiang, https://toomore.net/ <toomore0929@gmail.com>

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the “Software”), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED “AS IS”, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

http://toomore.mit-license.org/


CHANGELOG.md
# Changelog

### v1.4 (2017/07/08)
* cmd using spf13/cobra

### v1.3 (2017/07/07)
* Fixed twse change source link. #4

## v1.2

### v1.2.0 (2017/04/25)
* Fixed `realtime` #1 #3

### v1.2.1 (2017/04/26)
* Pass realtime testing
* Put gogrs:mini into test

## v1.1

### v1.1.0 (2017/04/07)
* Fixed `GetOSRamdiskPath`
* Add `Weight`, `WeightVolume`
* Add more test case

### v1.1.1 (2017/04/08)
* Fixed Dockerfile

### v1.1.2 (2017/04/08)
* Docker image using `alpine`
* Fixed image:`gogrs:mini` (`sh ./docker-mini/make.sh`)

### v1.1.3 (2017/04/10)
* Fixed Dockerfile
* Update open time 2011 ~ 2017

## v1.0.0 (2016/10/08)
* Stable release for dependency tools.

## v0.1.0 (2016/10/08)
* gogrs first package version.


twselist.go
package gogrs

import (
	"encoding/csv"
	"errors"
	"fmt"
	"io/ioutil"
	"net/http"
	"strings"
	"time"

	iconv "github.com/djimenez/iconv-go"
)

// TWSECLASS is a class list of TWSE.
var TWSECLASS = map[string]string{
	"0049":       "封閉式基金",
	"0099P":      "ETF",
	"01":         "水泥工業",
	"019919T":    "受益證券",
	"02":         "食品工業",
	"03":         "塑膠工業",
	"04":         "紡織纖維",
	"05":         "電機機械",
	"06":         "電器電纜",
	"07":         "化學生技醫療",
	"08":         "玻璃陶瓷",
	"09":         "造紙工業",
	"0999":       "認購權證", //(不含牛證)
	"0999B":      "熊證",
	"0999C":      "牛證",
	"0999G9":     "認股權憑證",
	"0999GA":     "附認股權特別股",
	"0999GD":     "附認股權公司債",
	"0999P":      "認售權證", //(不含熊證)
	"0999X":      "可展延牛證",
	"0999Y":      "可展延熊證",
	"10":         "鋼鐵工業",
	"11":         "橡膠工業",
	"12":         "汽車工業",
	"13":         "電子工業",
	"14":         "建材營造",
	"15":         "航運業",
	"16":         "觀光事業",
	"17":         "金融保險",
	"18":         "貿易百貨",
	"19":         "綜合",
	"20":         "其他",
	"21":         "化學工業",
	"22":         "生技醫療業",
	"23":         "油電燃氣業",
	"24":         "半導體業",
	"25":         "電腦及週邊設備業",
	"26":         "光電業",
	"27":         "通信網路業",
	"28":         "電子零組件業",
	"29":         "電子通路業",
	"30":         "資訊服務業",
	"31":         "其他電子業",
	"9299":       "存託憑證",
	"ALL":        "全部",
	"ALLBUT0999": "全部(不含權證、牛熊證、可展延牛熊證)",
	"CB":         "可轉換公司債",
}

// TWSEList is to get TWSE list.
type TWSEList struct {
	Date time.Time
}

// URL is to render urlpath.
func (l TWSEList) URL(strNo string) string {
	year, month, day := l.Date.Date()
	return fmt.Sprintf("%s%s", TWSEHOST, fmt.Sprintf(TWSELISTCSV, year, month, year, month, day, strNo))
}

// GetData is to get csv data.
func (l TWSEList) GetData(strNo string) ([][]string, error) {
	data, err := http.Get(l.URL(strNo))
	if err != nil {
		return nil, fmt.Errorf("Network fail: %s", err)
	}
	defer data.Body.Close()
	dataContentBig5, _ := ioutil.ReadAll(data.Body)
	dataContent, _ := iconv.ConvertString(string(dataContentBig5), "big5", "utf-8")
	csvArrayContent := strings.Split(dataContent, "\n")
	if len(csvArrayContent) > 5 {
		//for i := range csvArrayContent {
		//	fmt.Println(i, csvArrayContent[i])
		//}
		csvReader := csv.NewReader(strings.NewReader(strings.Join(csvArrayContent[2:len(csvArrayContent)-5], "\n")))
		return csvReader.ReadAll()
	}
	return nil, errors.New("Not enough data.")
}
